lib/datadog/appsec/utils/read_write_lock.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Datadog
+  module AppSec
+    module Utils
+      # Simple implentation for a fair read write lock
+      # It allows for as many concurrent readers but only one concurrent writer.
+      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
+      # no more reader threads can acquire the lock.
+      # The new reader threads will wait for the writer thread to finish.
+      class ReadWriteLock
+        def initialize
+          @reader_mutex    = Mutex.new
+          @reader_q        = ConditionVariable.new
+          @reader_count    = 0
+          @reader_releases = 0
+
+          @writer_mutex    = Mutex.new
+          @writer_q        = ConditionVariable.new
+          @writter         = false
+        end
+
+        def with_rlock
+          rlock
+          yield
+        ensure
+          runlock
+        end
+
+        def with_lock
+          lock
+          yield
+        ensure
+          unlock
+        end
+
+        def rlock
+          @reader_mutex.synchronize do
+            @reader_count += 1
+
+            writter = false
+
+            @writer_mutex.synchronize do
+              writter = @writter
+            end
+
+            @reader_q.wait(@reader_mutex) if writter
+          end
+        end
+
+        def runlock
+          @reader_mutex.synchronize do
+            @reader_releases += 1
+
+            @writer_mutex.synchronize { @writer_q.signal } if @reader_releases == @reader_count
+          end
+        end
```

Comment:
I'm assuming the reason for having `@reader_mutex` and `@writer_mutex` be separate is performance. But... are we really getting better performance out of this design?

It appears to me that the reader lock/unlock code paths are the most performance-sensitive: they're the ones that will get executed the most often, and that impact application code.

But these same paths have to acquire **both** locks! So I think we're actually penalizing them by having them acquire two locks most of the times.

My suggestion is going with a simpler 1-mutex implementation or at least change the 2-mutex implementation so the extra work is on the writers side, not on the readers.

Reply:
I concur, ideally `rlock` and `runlock` should be lockless but that's not possible in pure Ruby. The nested locks are scary.

As for performance, ultimately I would be only convinced with benchmarks of various kinds to see if it is performing better, and in which cases, so that we can understand the tradeoffs.

Reply:
@ivoanjo @lloeki 

I updated the benchmark so we can compare the actual implementation with a more straightforward implementation using a single mutex. 

The code for a single `Mutex`, is just that a single mutex, no counter no conditional variables. 

Here is the code of the benchmark 

```ruby
# frozen_string_literal: true

require_relative 'read_write_mutex'

require 'benchmark'

TOTAL_THREADS = 100 # set as high as practicable, for a better test

def test(lock)
  puts "READ INTENSIVE (80% read, 20% write):"
  single_test(lock, (TOTAL_THREADS * 0.8).floor, (TOTAL_THREADS * 0.2).floor)
  puts "WRITE INTENSIVE (80% write, 20% read):"
  single_test(lock, (TOTAL_THREADS * 0.2).floor, (TOTAL_THREADS * 0.8).floor)
  puts "BALANCED (50% read, 50% write):"
  single_test(lock, (TOTAL_THREADS * 0.5).floor, (TOTAL_THREADS * 0.5).floor)
end

def single_test(lock, n_readers, n_writers, reader_iterations=50, writer_iterations=50, reader_sleep=0.001, writer_sleep=0.001)
  puts "Testing #{lock.class} with #{n_readers} readers and #{n_writers} writers. Readers iterate #{reader_iterations} times, sleeping #{reader_sleep}s each time, writers iterate #{writer_iterations} times, sleeping #{writer_sleep}s each time"
  mutex = Mutex.new
  bad   = false
  data  = 0

  result = Benchmark.measure do
    readers = n_readers.times.collect do |i|
                Thread.new do
                  reader_iterations.times do
                    lock.with_rlock do
                      mutex.synchronize { bad = true } if (data % 2) != 0
                      sleep(reader_sleep)
                      mutex.synchronize { bad = true } if (data % 2) != 0
                    end
                  end
                end
              end
    writers = n_writers.times.collect do |i|
                Thread.new do
                  writer_iterations.times do
                    lock.with_lock do
                      value = (data += 1)
                      sleep(writer_sleep)
                      data  = value+1
                    end
                  end
                end
              end

    readers.each { |t| t.join }
    writers.each { |t| t.join }
    puts "BAD!!! Readers+writers overlapped!" if mutex.synchronize { bad }
    puts "BAD!!! Writers overlapped! Expected data #{(n_writers * writer_iterations * 2)} got #{data}" if data != (n_writers * writer_iterations * 2)
    puts "Expected data #{(n_writers * writer_iterations * 2)} got #{data}"
  end
  puts result
end

# for performance comparison with ReadWriteLock
class SimpleMutex
  def initialize; @mutex = Mutex.new; end
  def with_rlock
    @mutex.synchronize { yield }
  end
  alias :with_lock :with_rlock
end

test(SimpleMutex.new)
test(Locks::ReadWriteMutex.new)
```

The result of running it is: 

```
READ INTENSIVE (80% read, 20% write):
Testing SimpleMutex with 80 readers and 20 writers. Readers iterate 50 times, sleeping 0.001s each time, writers iterate 50 times, sleeping 0.001s each time
Expected data 2000 got 2000
  0.062491   0.126706   0.189197 (  6.403470)
WRITE INTENSIVE (80% write, 20% read):
Testing SimpleMutex with 20 readers and 80 writers. Readers iterate 50 times, sleeping 0.001s each time, writers iterate 50 times, sleeping 0.001s each time
Expected data 8000 got 8000
  0.059176   0.115606   0.174782 (  6.465615)
BALANCED (50% read, 50% write):
Testing SimpleMutex with 50 readers and 50 writers. Readers iterate 50 times, sleeping 0.001s each time, writers iterate 50 times, sleeping 0.001s each time
Expected data 5000 got 5000
  0.066263   0.132126   0.198389 (  6.446812)


READ INTENSIVE (80% read, 20% write):
Testing Locks::ReadWriteMutex with 80 readers and 20 writers. Readers iterate 50 times, sleeping 0.001s each time, writers iterate 50 times, sleeping 0.001s each time
Expected data 2000 got 2000
  0.040352   0.142648   0.183000 (  1.363321)
WRITE INTENSIVE (80% write, 20% read):
Testing Locks::ReadWriteMutex with 20 readers and 80 writers. Readers iterate 50 times, sleeping 0.001s each time, writers iterate 50 times, sleeping 0.001s each time
Expected data 8000 got 8000
  0.076468   0.122011   0.198479 (  5.293635)
BALANCED (50% read, 50% write):
Testing Locks::ReadWriteMutex with 50 readers and 50 writers. Readers iterate 50 times, sleeping 0.001s each time, writers iterate 50 times, sleeping 0.001s each time
Expected data 5000 got 5000
  0.052424   0.122417   0.174841 (  3.277331)
```

Reply:
Note that I'm not advocating here to use a single `Mutex` instead of a `ReadWriteMutex` (although that may be an interesting option to explore). I'm advocating to use a single `Mutex` in the underlying implementation of the `ReadWriteMutex`, instead of two different `Mutexes`, which makes implementation and validation more complex.